File: macos/Tests/WordingKitTests/PackIndexReaderTests.swift
import Foundation
import Testing

@testable import WordingKit

@Suite struct PackIndexReaderTests {
    static func data(_ json: String) -> Data { Data(json.utf8) }

    static let valid = """
        {
          "version": 1,
          "packs": [
            { "id": "spanish-travel", "name": "English → Spanish, travel",
              "description": "Airport words", "kind": "vocabulary", "wordCount": 25 },
            { "id": "it-interview-concepts", "name": "Backend interview concepts",
              "kind": "concepts", "wordCount": 44 }
          ]
        }
        """

    @Test func returnsEveryUsableEntry() throws {
        let entries = try PackIndexReader.read(Self.data(Self.valid))

        #expect(entries.count == 2)
        #expect(entries[0].id == "spanish-travel")
        #expect(entries[0].name == "English → Spanish, travel")
        #expect(entries[0].description == "Airport words")
        #expect(entries[0].wordCount == 25)
        #expect(entries[1].kind == PackKind.concepts)
    }

    @Test(arguments: ["../../words", "/etc/passwd", "https://elsewhere.example/evil", "con", ""])
    func dropsAnEntryWhoseIdentifierCouldChooseItsOwnAddress(id: String) throws {
        // The identifier is the whole of what decides the URL the app will fetch, so a
        // catalogue cannot smuggle one in.
        let json = """
            { "version": 1, "packs": [
              { "id": "\(id)", "name": "Bad", "wordCount": 1 },
              { "id": "spanish-travel", "name": "Good", "wordCount": 25 }] }
            """

        let entries = try PackIndexReader.read(Self.data(json))

        #expect(entries.count == 1)
        #expect(entries[0].id == "spanish-travel")
    }

    @Test func keepsTheGoodRowsWhenOneIsUnusable() throws {
        // A single bad entry must not hide the whole catalogue: it is the only way most
        // people will ever find a pack.
        let json = """
            { "version": 1, "packs": [
              { "id": "no-name", "name": "  ", "wordCount": 3 },
              { "id": "spanish-travel", "name": "Good", "wordCount": 25 }] }
            """

        let entries = try PackIndexReader.read(Self.data(json))

        #expect(entries.count == 1)
        #expect(entries[0].id == "spanish-travel")
    }

    @Test func keepsOnlyTheFirstOfADuplicatedIdentifier() throws {
        let json = """
            { "version": 1, "packs": [
              { "id": "spanish-travel", "name": "First", "wordCount": 1 },
              { "id": "Spanish-Travel", "name": "Second", "wordCount": 2 }] }
            """

        let entries = try PackIndexReader.read(Self.data(json))

        #expect(entries.count == 1)
        #expect(entries[0].name == "First")
    }

    @Test func tidiesTextThatWouldBreakTheList() throws {
        let json = """
            { "version": 1, "packs": [
              { "id": "x", "name": "two\\nlines ", "description": "\\ttabbed", "wordCount": -5 }] }
            """

        let entry = try #require(try PackIndexReader.read(Self.data(json)).first)

        #expect(entry.name == "two lines")
        #expect(entry.description == "tabbed")
        #expect(entry.wordCount == 0)
    }

    @Test func defaultsAMissingKindToVocabulary() throws {
        let json = #"{ "version": 1, "packs": [{ "id": "x", "name": "X", "wordCount": 1 }] }"#

        #expect(try PackIndexReader.read(Self.data(json)).first?.kind == PackKind.vocabulary)
    }

    @Test func stopsAtTheEntryLimit() throws {
        let rows = (0..<(PackLimits.maxIndexEntries + 50))
            .map { #"{ "id": "pack-\#($0)", "name": "P\#($0)", "wordCount": 1 }"# }
            .joined(separator: ",")

        let entries = try PackIndexReader.read(Self.data(#"{ "version": 1, "packs": [\#(rows)] }"#))

        #expect(entries.count == PackLimits.maxIndexEntries)
    }

    @Test func anEmptyCatalogueIsNotAnError() throws {
        // Nothing published yet is a state, not a failure.
        #expect(try PackIndexReader.read(Self.data(#"{ "version": 1, "packs": [] }"#)).isEmpty)
    }

    @Test(arguments: ["not json", "[]"])
    func refusesWhatIsNotACatalogue(json: String) {
        #expect(throws: WordPackError.self) { try PackIndexReader.read(Self.data(json)) }
    }

    @Test func refusesAPayloadOverTheLimit() {
        let padding = String(repeating: "x", count: PackLimits.maxPayloadBytes + 1)
        let json = #"{ "version": 1, "packs": [{ "id": "x", "name": "\#(padding)", "wordCount": 1 }] }"#

        #expect(throws: WordPackError.tooLarge) { try PackIndexReader.read(Self.data(json)) }
    }

    @Test func packUrlIsBuiltFromTheIndexAddressAndTheIdentifierAlone() throws {
        let index = URL(string: "https://example.com/data/index.json")!

        #expect(
            try PackSource.packURL(index: index, id: "spanish-travel").absoluteString
                == "https://example.com/data/spanish-travel.json")

        // A mirror serves the same catalogue without a single address being rewritten.
        #expect(
            try PackSource.packURL(
                index: URL(string: "https://mirror.example.org/wording/index.json")!,
                id: "spanish-travel"
            ).absoluteString == "https://mirror.example.org/wording/spanish-travel.json")
    }

    @Test func packUrlRefusesAnIdentifierThatIsNotASafeSlug() {
        let index = URL(string: "https://example.com/data/index.json")!

        #expect(throws: WordPackError.unsafeId) {
            try PackSource.packURL(index: index, id: "../../secrets")
        }
    }
}

File: tests/Wording.Core.Tests/PackIndexReaderTests.cs
using System.Text;
using Wording.Core.Packs;

namespace Wording.Core.Tests;

public class PackIndexReaderTests
{
    static byte[] Bytes(string json) => Encoding.UTF8.GetBytes(json);

    const string Valid = """
        {
          "version": 1,
          "packs": [
            { "id": "spanish-travel", "name": "English → Spanish, travel",
              "description": "Airport words", "kind": "vocabulary", "wordCount": 25 },
            { "id": "it-interview-concepts", "name": "Backend interview concepts",
              "kind": "concepts", "wordCount": 44 }
          ]
        }
        """;

    [Fact]
    public void Read_ReturnsEveryUsableEntry()
    {
        var entries = PackIndexReader.Read(Bytes(Valid));

        Assert.Equal(2, entries.Count);
        Assert.Equal("spanish-travel", entries[0].Id);
        Assert.Equal("English → Spanish, travel", entries[0].Name);
        Assert.Equal("Airport words", entries[0].Description);
        Assert.Equal(25, entries[0].WordCount);
        Assert.Equal(PackKind.Concepts, entries[1].Kind);
    }

    [Theory]
    [InlineData("../../words")]
    [InlineData("/etc/passwd")]
    [InlineData("https://elsewhere.example/evil")]
    [InlineData("con")]
    [InlineData("")]
    public void Read_DropsAnEntryWhoseIdentifierCouldChooseItsOwnAddress(string id)
    {
        // The identifier is the whole of what decides the URL the app will fetch, so a
        // catalogue cannot smuggle one in.
        var json = $$"""
            { "version": 1, "packs": [
              { "id": "{{id}}", "name": "Bad", "wordCount": 1 },
              { "id": "spanish-travel", "name": "Good", "wordCount": 25 }] }
            """;

        var entries = PackIndexReader.Read(Bytes(json));

        Assert.Equal("spanish-travel", Assert.Single(entries).Id);
    }

    [Fact]
    public void Read_KeepsTheGoodRowsWhenOneIsUnusable()
    {
        // A single bad entry must not hide the whole catalogue: it is the only way most
        // people will ever find a pack.
        var json = """
            { "version": 1, "packs": [
              { "id": "no-name", "name": "  ", "wordCount": 3 },
              { "id": "spanish-travel", "name": "Good", "wordCount": 25 }] }
            """;

        Assert.Equal("spanish-travel", Assert.Single(PackIndexReader.Read(Bytes(json))).Id);
    }

    [Fact]
    public void Read_KeepsOnlyTheFirstOfADuplicatedIdentifier()
    {
        var json = """
            { "version": 1, "packs": [
              { "id": "spanish-travel", "name": "First", "wordCount": 1 },
              { "id": "Spanish-Travel", "name": "Second", "wordCount": 2 }] }
            """;

        Assert.Equal("First", Assert.Single(PackIndexReader.Read(Bytes(json))).Name);
    }

    [Fact]
    public void Read_TidiesTextThatWouldBreakTheList()
    {
        var json = """
            { "version": 1, "packs": [
              { "id": "x", "name": "two\nlines ", "description": "\ttabbed", "wordCount": -5 }] }
            """;

        var entry = Assert.Single(PackIndexReader.Read(Bytes(json)));

        Assert.Equal("two lines", entry.Name);
        Assert.Equal("tabbed", entry.Description);
        Assert.Equal(0, entry.WordCount);
    }

    [Fact]
    public void Read_DefaultsAMissingKindToVocabulary()
    {
        var json = """{ "version": 1, "packs": [{ "id": "x", "name": "X", "wordCount": 1 }] }""";

        Assert.Equal(PackKind.Vocabulary, Assert.Single(PackIndexReader.Read(Bytes(json))).Kind);
    }

    [Fact]
    public void Read_StopsAtTheEntryLimit()
    {
        var rows = string.Join(",", Enumerable
            .Range(0, PackLimits.MaxIndexEntries + 50)
            .Select(index => $$"""{ "id": "pack-{{index}}", "name": "P{{index}}", "wordCount": 1 }"""));

        Assert.Equal(
            PackLimits.MaxIndexEntries,
            PackIndexReader.Read(Bytes($$"""{ "version": 1, "packs": [{{rows}}] }""")).Count);
    }

    [Fact]
    public void Read_AnEmptyCatalogueIsNotAnError()
    {
        // Nothing published yet is a state, not a failure.
        Assert.Empty(PackIndexReader.Read(Bytes("""{ "version": 1, "packs": [] }""")));
    }

    [Theory]
    [InlineData("not json")]
    [InlineData("[]")]
    public void Read_RefusesWhatIsNotACatalogue(string json)
    {
        Assert.Equal(
            PackProblem.Malformed,
            Assert.Throws<WordPackException>(() => PackIndexReader.Read(Bytes(json))).Problem);
    }

    [Fact]
    public void Read_RefusesAPayloadOverTheLimit()
    {
        var padding = new string('x', PackLimits.MaxPayloadBytes + 1);
        var json = $$"""{ "version": 1, "packs": [{ "id": "x", "name": "{{padding}}", "wordCount": 1 }] }""";

        Assert.Equal(
            PackProblem.TooLarge,
            Assert.Throws<WordPackException>(() => PackIndexReader.Read(Bytes(json))).Problem);
    }

    [Fact]
    public void PackUrl_IsBuiltFromTheIndexAddressAndTheIdentifierAlone()
    {
        var index = new Uri("https://example.com/data/index.json");

        Assert.Equal("https://example.com/data/spanish-travel.json", PackSource.PackUrl(index, "spanish-travel").ToString());

        // A mirror serves the same catalogue without a single address being rewritten.
        Assert.Equal(
            "https://mirror.example.org/wording/spanish-travel.json",
            PackSource.PackUrl(new Uri("https://mirror.example.org/wording/index.json"), "spanish-travel").ToString());
    }

    [Fact]
    public void PackUrl_RefusesAnIdentifierThatIsNotASafeSlug()
    {
        var index = new Uri("https://example.com/data/index.json");

        Assert.Equal(
            PackProblem.UnsafeId,
            Assert.Throws<WordPackException>(() => PackSource.PackUrl(index, "../../secrets")).Problem);
    }
}
